File: practice/3_Longest_Substring_Without_Repeating_Characters.js
function practice(s) {
    let map = new Map(),
        max = 0,
        i = 0,
        j = 0;

    while (j < s.length) {
        map.set(s[j], map.get(s[j]) + 1 || 1);

        if (map.get(s[j]) > 1) {
            while (map.get(s[j]) > 1) {
                map.set(s[i], map.get(s[i]) - 1);

                if (map.get(s[i]) <= 0) {
                    map.delete(s[i]);
                }
                i++;
            }
            j++;
        } else {
            max = Math.max(max, j - i + 1);
            j++;
        }
    }

    return max;

    console.log(max);
}

console.log(practice('abcabcbb'));
// console.log(practice('bbbbb'));
// console.log(practice('pwwkew'));

File: practice/3_Longest_Substring_Without_Repeating_Characters.kt
fun main() {
	practice("abcabcbb")
}

fun practice(s: String): Int {
	val map = mutableMapOf<Char, Int>();
	var max = 0;
	var i = 0;
	var j = 0;

	while(j < s.length) {
		val charJ = s[j];
		map[charJ] = map.getOrDefault(charJ, 0) + 1

		if(map[charJ]!! > 1) {

			while(map[charJ]!! > 1) {
				val charI = s[i];
				map[charI] = map[charI]!! - 1

				if(map[charI]!! <= 0) {
					map.remove(charI)
				}
				i++;
			}
			j++
		} else {
			max = maxOf(max, j-i+1)
			j++
		}
	}

	return max
}
